### PR TITLE
Split default parameter extractors and interceptor factories in to their own modules

### PR DESCRIPTION
**Pull request by JGulbronson**

### CODE REVIEW

File: exemplar/src/main/java/com/squareup/exemplar/ExemplarJavaApp.java [hunk 1]
```diff
@@ -5,16 +5,14 @@
 import misk.environment.EnvironmentModule;
 import misk.hibernate.HibernateModule;
 import misk.moshi.MoshiModule;
-import misk.web.WebActionsModule;
-import misk.web.jetty.JettyModule;
+import misk.web.jetty.WebModule;
```
Comment (maniksurtani) [inline exemplar/src/main/java/com/squareup/exemplar/ExemplarJavaApp.java]: `misk.web.WebModule`?

Comment (maniksurtani) [reply exemplar/src/main/java/com/squareup/exemplar/ExemplarJavaApp.java]: Perhaps maybe then it is best to also have a `misk.web.jetty.JettyModule` that is installed by `misk.web.WebModule`...

Comment (JGulbronson) [reply exemplar/src/main/java/com/squareup/exemplar/ExemplarJavaApp.java]: Good call, fixed.

File: misk/src/main/kotlin/misk/MiskDefault.kt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package misk
+
+import com.google.inject.BindingAnnotation
+import kotlin.annotation.AnnotationRetention.RUNTIME
+
+@BindingAnnotation
+@Retention(RUNTIME)
+@Target(AnnotationTarget.CLASS, AnnotationTarget.VALUE_PARAMETER, AnnotationTarget.TYPE)
+annotation class MiskDefault
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/MiskDefault.kt]: hmmm. . . without Javadoc I have no idea what this does

File: misk/src/main/kotlin/misk/inject/Guice.kt [hunk 3]
```diff
@@ -12,12 +12,21 @@ import javax.inject.Singleton
...
 
-@Suppress("UNCHECKED_CAST")
 internal inline fun <reified T : Any> Binder.addMultibinderBinding(): LinkedBindingBuilder<T> {
+    return addMultibinderBinding(null)
+}
+
+internal inline fun <reified T : Any, reified A : Annotation> Binder.addMultibinderBindingWithAnnotation() = addMultibinderBinding<T>(A::class.java)
+
+@Suppress("UNCHECKED_CAST")
+private inline fun <reified T : Any> Binder.addMultibinderBinding(annotation: Class<out Annotation>?): LinkedBindingBuilder<T> {
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/inject/Guice.kt]: can you make `annotation` default to null and save some overloads?

File: misk/src/main/kotlin/misk/web/WebActionModule.kt [hunk 4]
```diff
@@ -26,7 +27,11 @@ class WebActionModule<A : WebAction> private constructor(
         val provider = getProvider(webActionClass.java)
         @Suppress("UNCHECKED_CAST")
         val binder: Multibinder<BoundAction<A, *>> = Multibinder.newSetBinder(
-                binder(), TypeLiteral.get(Types.newParameterizedType(BoundAction::class.java, Types.subtypeOf(WebAction::class.java), Types.subtypeOf(Object::class.java)))) as Multibinder<BoundAction<A, *>>
+                binder(),
+                TypeLiteral.get(Types.newParameterizedType(BoundAction::class.java,
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/WebActionModule.kt]: we should make some Kotlin sugar for these

Comment (swankjesse) [reply misk/src/main/kotlin/misk/web/WebActionModule.kt]:
```
  parameterizedType<BoundAction>(subtypeOf<WebAction>, subtypeOf<Object>).typeLiteral();
```

File: misk/src/main/kotlin/misk/web/WebActionModule.kt [hunk 5]
```diff
@@ -67,14 +72,17 @@ internal class BoundActionProvider<A : WebAction, R>(
     val httpMethod: HttpMethod
 ) : Provider<BoundAction<A, *>> {
 
-    @Inject lateinit var interceptorFactories: List<Interceptor.Factory>
+    @Inject lateinit var userProvidedInterceptorFactories: List<Interceptor.Factory>
+    @Inject lateinit var defaultInterceptorFactories: @MiskDefault List<Interceptor.Factory>
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/WebActionModule.kt]: I don’t think I like this name

File: misk/src/main/kotlin/misk/web/WebActionModule.kt [hunk 6]
```diff
@@ -67,14 +72,17 @@ internal class BoundActionProvider<A : WebAction, R>(
...
     @Inject lateinit var parameterExtractorFactories: List<ParameterExtractor.Factory>
 
     override fun get(): BoundAction<A, *> {
         val action = function.asAction()
 
         val interceptors = ArrayList<Interceptor>()
-        interceptorFactories.mapNotNullTo(interceptors) { it.create(action) }
+        // Ensure that default interceptors are called before any user provided interceptors
+        defaultInterceptorFactories.mapNotNullTo(interceptors) { it.create(action) }
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/WebActionModule.kt]: I find `mapNotNullTo` to be a very awkward method

File: misk/src/main/kotlin/misk/web/extractors/DefaultParameterExtractorsModule.kt [hunk 7]
```diff
@@ -0,0 +1,13 @@
+package misk.web.extractors
+
+import com.google.inject.AbstractModule
+import misk.inject.addMultibinderBinding
+import misk.inject.to
+
+class DefaultParameterExtractorsModule : AbstractModule() {
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/extractors/DefaultParameterExtractorsModule.kt]: Would this ever be used standalone? Why not just put this in WebModule ?

File: misk/src/main/kotlin/misk/web/extractors/JsonBodyParameterExtractorFactory.kt [hunk 8]
```diff
@@ -1,25 +1,25 @@
...
-import com.squareup.moshi.Moshi
 import java.util.regex.Matcher
 import javax.inject.Inject
 import javax.inject.Singleton
 import kotlin.reflect.KParameter
 import kotlin.reflect.full.findAnnotation
 import kotlin.reflect.jvm.javaType
 
+/**
+ * Creates a [ParameterExtractor] that returns the JSON body of a request if the parameter
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/extractors/JsonBodyParameterExtractorFactory.kt]: Nice

File: misk/src/main/kotlin/misk/web/interceptors/DefaultInterceptorsModule.kt [hunk 9]
```diff
@@ -0,0 +1,22 @@
+package misk.web.interceptors
+
+import com.google.inject.AbstractModule
+import com.google.inject.multibindings.Multibinder
+import misk.Interceptor
+import misk.MiskDefault
+import misk.inject.addMultibinderBindingWithAnnotation
+import misk.inject.to
+
+class DefaultInterceptorsModule : AbstractModule() {
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/interceptors/DefaultInterceptorsModule.kt]: Same. Does this need to be its own module?

File: misk/src/main/kotlin/misk/inject/Guice.kt (hunk 3 above)
Comment (JGulbronson) [reply misk/src/main/kotlin/misk/inject/Guice.kt]: Right, that's the Java coming through when I forget about default params.

File: misk/src/main/kotlin/misk/web/WebActionModule.kt (hunk 4 above)
Comment (JGulbronson) [reply misk/src/main/kotlin/misk/web/WebActionModule.kt]: Agreed, I think that's the last "ugly" usage we haven't cleaned up yet. I'll follow up with that in a separate PR tomorrow.

File: misk/src/main/kotlin/misk/web/WebActionModule.kt (hunk 5 above)
Comment (JGulbronson) [reply misk/src/main/kotlin/misk/web/WebActionModule.kt]: `miskInterceptorFactories`? I think we'll end up with a few layers of Interceptors like OkHTTP, so we'll probably have to play around with it then.

File: misk/src/main/kotlin/misk/web/WebActionModule.kt (hunk 6 above)
Comment (JGulbronson) [reply misk/src/main/kotlin/misk/web/WebActionModule.kt]: Yeah, it doesn't really map well coming from other languages. I'll see if I can make it more grok-able.

File: misk/src/main/kotlin/misk/web/extractors/DefaultParameterExtractorsModule.kt (hunk 7 above)
Comment (JGulbronson) [reply misk/src/main/kotlin/misk/web/extractors/DefaultParameterExtractorsModule.kt]: Yeah I'll move it to `WebModule`. After talking to @maniksurtani I moved some modules around, and these don't make sense on their own anymore.

**Pull request merged**